sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -26,51 +26,59 @@ namespace sycl {
 __SYCL_INLINE_VER_NAMESPACE(_V1) {
 namespace detail {
 
-void Scheduler::waitForRecordToFinish(MemObjRecord *Record,
-                                      ReadLockT &GraphReadLock) {
-#ifdef XPTI_ENABLE_INSTRUMENTATION
-  // Will contain the list of dependencies for the Release Command
-  std::set<Command *> DepCommands;
-#endif
+inline bool Scheduler::checkLeavesCompletion(LeavesCollection &Leaves,
+                                             ReadLockT &GraphReadLock,
+                                             bool Blocking) {
   std::vector<Command *> ToCleanUp;
-  for (Command *Cmd : Record->MReadLeaves) {
-    EnqueueResultT Res;
-    bool Enqueued = GraphProcessor::enqueueCommand(Cmd, Res, ToCleanUp);
-    if (!Enqueued && EnqueueResultT::SyclEnqueueFailed == Res.MResult)
-      throw runtime_error("Enqueue process failed.",
-                          PI_ERROR_INVALID_OPERATION);
-#ifdef XPTI_ENABLE_INSTRUMENTATION
-    // Capture the dependencies
-    DepCommands.insert(Cmd);
-#endif
-    GraphProcessor::waitForEvent(Cmd->getEvent(), GraphReadLock, ToCleanUp);
-  }
-  for (Command *Cmd : Record->MWriteLeaves) {
-    EnqueueResultT Res;
-    bool Enqueued = GraphProcessor::enqueueCommand(Cmd, Res, ToCleanUp);
-    if (!Enqueued && EnqueueResultT::SyclEnqueueFailed == Res.MResult)
-      throw runtime_error("Enqueue process failed.",
-                          PI_ERROR_INVALID_OPERATION);
-#ifdef XPTI_ENABLE_INSTRUMENTATION
-    DepCommands.insert(Cmd);
-#endif
-    GraphProcessor::waitForEvent(Cmd->getEvent(), GraphReadLock, ToCleanUp);
+  for (Command *Cmd : Leaves) {
+    // Expect that mem object release is not responsible for dependency commands
+    // enqueue
+    if (Cmd->getEvent()->isCompleted())
+      continue;
+
+    if (Blocking)
+      GraphProcessor::waitForEvent(Cmd->getEvent(), GraphReadLock, ToCleanUp);
+    else
+      return false;
   }
+  return true;
+}
+
+bool Scheduler::waitForRecordToFinish(MemObjRecord *Record,
+                                      ReadLockT &GraphReadLock, bool Blocking) {
+  assert(Record);
+  if (!checkLeavesCompletion(Record->MReadLeaves, GraphReadLock, Blocking))
+    return false;
+  if (!checkLeavesCompletion(Record->MWriteLeaves, GraphReadLock, Blocking))
+    return false;
```

Comment:
Can we avoid modification of `waitForRecordToFinish` and just directly call `checkLeavesCompletion` from `cleanupDeferredMemObjects`?

Reply:
fixed in https://github.com/intel/llvm/pull/6837/commits/106132246e56ca1834ad8607ffbf15b937954d7b